onnxruntime/core/providers/coreml/builders/impl/split_op_builder.cc

```diff
@@ -0,0 +1,151 @@
+
+#if defined(__APPLE__)
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+
+namespace onnxruntime {
+namespace coreml {
+
+class SplitOpBuilder : public BaseOpBuilder {
+  // Add operator related
+#ifdef __APPLE__
+ private:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override;
+#endif
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+
+  // Split opset 13- uses "split" as attribute. Currently it's not supported.
+  int GetMinSupportedOpSet(const Node& /* node */) const override { return 13; }
+};
+
+// Add operator related
+
+#ifdef __APPLE__
+
+void SplitOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  const auto& input_defs = node.InputDefs();
+
+  if (input_defs.size() > 1 && input_defs[1]->Exists()) {  // optional second input "split"
+    model_builder.AddInitializerToSkip(input_defs[1]->Name());
+  }
+}
+
+Status SplitOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                             const Node& node,
+                                             const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  std::vector<int64_t> data_shape;
+  ORT_RETURN_IF_NOT(GetStaticShape(*node.InputDefs()[0], data_shape, logger), "Failed to get input shape.");
+
+  NodeAttrHelper helper(node);
+  const auto axis = helper.Get("axis", 0);
+
+  // optional attribute introduced since opset 18
+  uint64_t num_outputs = 2;
+
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+  auto* coreml_splitnd = layer->mutable_splitnd();
+  coreml_splitnd->set_axis(axis);
+
+  if (input_defs.size() > 1) {
+    const auto& split_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());
+    Initializer unpacked_tensor(split_tensor);
+    auto split_span = unpacked_tensor.DataAsSpan<int64_t>();
+    auto split_sizes = split_span.size();
+    num_outputs = SafeInt<uint64_t>(split_sizes);
+    // numSplits will be ignored in this case, but setting it here in the coreml layerparam.
+    coreml_splitnd->set_numsplits(num_outputs);
+    for (size_t i = 0; i < split_sizes; i++) {
+      coreml_splitnd->add_splitsizes(SafeInt<uint64_t>(split_span[i]));
+    }
+  } else if (node.SinceVersion() < 18) {
+    coreml_splitnd->set_numsplits(num_outputs);
+    auto input_dim_value_at_axis = data_shape[HandleNegativeAxis(axis, data_shape.size())];
+    coreml_splitnd->add_splitsizes(SafeInt<uint64_t>(input_dim_value_at_axis - input_dim_value_at_axis / 2));
+    coreml_splitnd->add_splitsizes(SafeInt<uint64_t>(input_dim_value_at_axis / 2));
```

Comment:
See above comment. I believe num_outputs should be based on the number of node outputs if not explicitly provided.

https://github.com/onnx/onnx/blob/main/docs/Operators.md#examples-153

The 'if' where the splits are explicitly provided seems fine.

If splits are not provided, does it diverge into these 2 branches?
- opset < 18 the splits are even, so we just need to set num_outputs
- opset >= 18, the splits may or may not be even
  - if even, can set num_outputs
  - if not, you need to calculate the split sizes. see cpu ep [implementation](https://github.com/microsoft/onnxruntime/blob/d9558857913b32a9d701d217362017b892a2769c/onnxruntime/core/providers/cpu/tensor/split.cc#L85-L93)

Reply:
right, thanks for pointing out this.

sorry - I think I had some misunderstanding of the spec. The branch here was lacking some logic.. 

updated now and used the existing split cpu impl  for calculating the split size, etc. re-validated that it passes unit test case.